cmd/kubeadm/app/apis/kubeadm/types.go

```diff
@@ -24,68 +24,149 @@ import (
-	// controller manager are managed by Kubernetes itself. This option is likely to
-	// become the default in the future.
-	SelfHosted bool
+	// AdvertiseAddress sets the address for the API server to advertise.
+	AdvertiseAddress string `json:"advertiseAddress"`
+	// BindPort sets the secure port for the API Server to bind to
+	BindPort int32 `json:"bindPort"`
+	// AuthorizationModes specifies which authorization modes to use for the api server
+	AuthorizationModes []string `json:"authorizationModes"`
+	// ServingCertExtraSANs sets extra Subject Alternative Names for the API Server signing cert
+	ServingCertExtraSANs []string `json:"servingCertExtraSANs"`
+}
 
-	APIServerExtraArgs         map[string]string
-	ControllerManagerExtraArgs map[string]string
-	SchedulerExtraArgs         map[string]string
+// ControllerManager specifies how the controller manager component is set up and configured
+type ControllerManager struct {
+	// Image specifies which container image to use for running the controller manager. If empty, automatically populated by kubeadm using the ImageRepository and KubernetesVersion
+	Image string `json:"image"`
+	// ExtraArgs lets the user customize and/or override the arguments that are given to the controller manager
+	ExtraArgs map[string]string `json:"extraArgs"`
+	// ExtraHostPathVolumes lets the user add new hostPath volumes and volumeMounts as needed
+	ExtraHostPathVolumes []HostPathVolume `json:"extraHostPathVolumes"`
+}
 
-	// APIServerCertSANs sets extra Subject Alternative Names for the API Server signing cert
-	APIServerCertSANs []string
-	// CertificatesDir specifies where to store or look for all required certificates
-	CertificatesDir string
+// Scheduler specifies how the scheduler component is set up and configured
+type Scheduler struct {
+	// Image specifies which container image to use for running the scheduler. If empty, automatically populated by kubeadm using the ImageRepository and KubernetesVersion
+	Image string `json:"image"`
+	// ExtraArgs lets the user customize and/or override the arguments that are given to the scheduler
+	ExtraArgs map[string]string `json:"extraArgs"`
+	// ExtraHostPathVolumes lets the user add new hostPath volumes and volumeMounts as needed
+	ExtraHostPathVolumes []HostPathVolume `json:"extraHostPathVolumes"`
+}
 
-	// ImageRepository what container registry to pull control plane images from
-	ImageRepository string
-	// UnifiedControlPlaneImage specifies if a specific container image should be used for all control plane components
-	UnifiedControlPlaneImage string
+// HostPathVolume is a high-level abstraction for a v1.Volume and v1.VolumeMount attached to the Pod the component is running in
+type HostPathVolume struct {
+	// Name specifies the name of the Volume and VolumeMount
+	Name string `json:"name"`
+	// HostPath specifies the path on the host (can be a file or a directory) that should be mounted into the container
```

Comment:
Because these comments might end up in documentation we avoid repeating the field name in the comment.